frontend/src/components/widgets/CustomComponent/ComponentInstance.test.tsx

```diff
@@ -0,0 +1,287 @@
+/**
+ * @license
+ * Copyright 2018-2020 Streamlit Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from "react"
+import { fromJS } from "immutable"
+import { mount, ReactWrapper } from "enzyme"
+import { ComponentRegistry } from "./ComponentRegistry"
+import {
+  ComponentInstance,
+  Props,
+  State,
+  StreamlitMessageType,
+  ComponentMessageType,
+} from "./ComponentInstance"
+import { Source, WidgetStateManager } from "lib/WidgetStateManager"
+import ErrorElement from "components/shared/ErrorElement"
+import { logWarning } from "lib/log"
+import { buildHttpUri } from "lib/UriUtil"
+
+// Mock log functions.
+jest.mock("lib/log")
+
+// Mock uri utils.
+jest.mock("lib/UriUtil")
+const mockedBuildHttpUri = buildHttpUri as jest.Mock
+mockedBuildHttpUri.mockImplementation(() => "registry/url")
+
+const getProps = (elementProps = {}): Props => ({
+  element: fromJS({
+    argsDataframe: [],
+    argsJson: '{"foo": "bar"}',
+    componentName: "some.component",
+    id: "some_id",
+    url: "some/url",
+    ...elementProps,
+  }),
+  registry: new ComponentRegistry(() =>
+    fromJS({
+      host: "streamlit.mock",
+      port: 80,
+      basePath: "",
+    })
+  ),
+  width: 100,
+  disabled: false,
+  widgetMgr: new WidgetStateManager(jest.fn()),
+})
+
+describe("ComponentInstance", () => {
+  let wrapper: ReactWrapper<Props, State, ComponentInstance>
+  let instance: ComponentInstance
+
+  beforeEach(() => {
+    const props = getProps()
+    const mountNode = document.createElement("div")
+    document.body.appendChild(mountNode)
+
+    // We need to mount the ComponentInstance to an existing DOM element.
+    // Otherwise, iframe contentWindow is not available.
+    wrapper = mount(<ComponentInstance {...props} />, { attachTo: mountNode })
+    instance = wrapper.instance()
+  })
+
+  describe("handleSetComponentValue", () => {
+    let handleSetComponentValue: (data: any, source: Source) => void
```

Comment:
Hm, so if I understand correctly we're trying to unit test each of these private instance methods of ComponentInstance, and to make it work in Typescript we're accessing and assigning them using reflection, more or less instance["methodName"].

I'm generally a bit skeptical about this kind of approach -- to me,

```
// @ts-ignore
instance.methodName(params)
```

reads much more fluently than

```
methodName = instance["methodName"]
methodName(params)
```

(and here it's even more complicated with the beforeEach setups.) WDYT?

Reply:
Agree. I'll also add some comments explaining why are we doing that.